File: bot.py
# ================= IMPORT =================
import os
import time
import random
import threading
from flask import Flask
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup, ReplyKeyboardMarkup
from telegram.ext import (
    ApplicationBuilder,
    CommandHandler,
    MessageHandler,
    ContextTypes,
    filters
)

# ================= CONFIG =================
# ⚠️ BOT_TOKEN langsung dari environment Choreo.dev, jangan pakai .env
BOT_TOKEN = os.environ.get("BOT_TOKEN")
if not BOT_TOKEN:
    raise ValueError("❌ BOT_TOKEN environment variable not found! Pastikan sudah set di Choreo.")

ADMIN_USER_ID = 8146896736
SOURCE_CHAT_ID = -1002626291566
TARGET_CHAT_IDS = [-1003175423118, -1003443785953]
AUTO_DELETE_DELAY = 2  # detik

# ================= SUBSCRIBERS (in-memory) =================
subscribers = set()

# ================= MENU DATA =================
menu_data = {
    "🌟 NEW REGISTER FREE 🌟": {
        "url": "https://afb88.hfcapital.top/",
        "media": "https://i.ibb.co/BK2LVQ6t/image.png",
        "caption": "🌟 NEW REGISTER BONUS 🌟\n⚠️ 1 NAMA 1 ID SAHAJA..."
    },
    "📘 SHARE FACEBOOK 📘": {
        "url": "https://afb88.hfcapital.top/",
        "media": "https://i.ibb.co/Z6B55VcX/image.png",
        "caption": "📘 SHARE FACEBOOK 📘\n🧧 FREE CREDIT RM68..."
    },
    "🔥 DAILY APPS FREE 🔥": {
        "url": "https://afb88.hfcapital.top/",
        "media": "https://i.ibb.co/nsmVQFbg/image.png",
        "caption": "🔥 DAILY APPS FREE 🔥\n🎁 Free Credit RM20..."
    },
    "🌞 SOCIAL MEDIA 🌞": {
        "url": "https://afb88.hfcapital.top/",
        "media": "https://i.ibb.co/HfyD6DWw/image.png",
        "caption": "🌞 SOCIAL MEDIA 🌞\n📌FOLLOW SOCIAL MEDIA..."
    },
    "🎉 TELEGRAM BONUS 🎉": {
        "url": "https://afb88.hfcapital.top/",
        "media": "https://i.ibb.co/21qTqmtY/image.png",
        "caption": "🎉 TELEGRAM BONUS 🎉\n🎁 SUBSCRIBE TELEGRAM BONUS..."
    }
}

# ================= HELPER FUNCTIONS =================
async def add_subscriber(user_id: int, username: str, app):
    if user_id not in subscribers:
        subscribers.add(user_id)
        try:
            await app.bot.send_message(
                chat_id=ADMIN_USER_ID,
                text=f"📌 Subscriber baru: {username} ({user_id})"
            )
        except Exception as e:
            print("Notif admin gagal:", e)

# ================= START / MENU =================
async def start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    user = update.effective_user
    username = f"@{user.username}" if user.username else user.first_name or "Bossku"

    await add_subscriber(user.id, username, context.application)

    inline_buttons = InlineKeyboardMarkup([
        [InlineKeyboardButton("📢 SUBSCRIBE CHANNEL", url="https://t.me/afb88my")],
        [InlineKeyboardButton("💬 GROUP CUCI & TIPS GAME", url="https://t.me/+b685QE242dMxOWE9")],
        [InlineKeyboardButton("🌐 REGISTER & LOGIN", url="https://afb88my1.com/")],
        [InlineKeyboardButton("🎁 GROUP HADIAH AFB88", url="https://t.me/Xamoi2688")]
    ])

    reply_keyboard = ReplyKeyboardMarkup([
        ["🌟 NEW REGISTER FREE 🌟"],
        ["📘 SHARE FACEBOOK 📘"],
        ["🔥 DAILY APPS FREE 🔥", "🌞 SOCIAL MEDIA 🌞"],
        ["🎉 TELEGRAM BONUS 🎉"]
    ], resize_keyboard=True)

    try:
        await update.message.reply_animation(
            "https://media.giphy.com/media/tXSLbuTIf37SjvE6QY/giphy.gif",
            caption=f"👋 Hi {username} Bossku 😘 Sila join semua group dulu ya ...",
            reply_markup=inline_buttons
        )
    except Exception:
        pass

    await update.message.reply_text("➤ CLICK /start TO BACK MENU", reply_markup=reply_keyboard)

# ================= MENU HANDLER =================
async def menu_handler(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if update.effective_chat.type != "private":
        return

    text = update.message.text
    data = menu_data.get(text)
    if not data:
        return

    try:
        await update.message.reply_photo(
            photo=data["media"],
            caption=data["caption"],
            reply_markup=InlineKeyboardMarkup([[InlineKeyboardButton("CLAIM 🎁", url=data["url"])]]),
        )
    except Exception as e:
        print("Menu reply error:", e)

# ================= FORWARD COMMAND =================
async def forward(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if update.effective_user.id != ADMIN_USER_ID or update.effective_chat.id != SOURCE_CHAT_ID:
        return

    reply_to = update.message.reply_to_message
    if not reply_to:
        return

    # Forward ke TARGET_CHAT_IDS
    for target_id in TARGET_CHAT_IDS:
        if target_id == SOURCE_CHAT_ID:
            continue
        try:
            await context.bot.forward_message(
                chat_id=target_id,
                from_chat_id=reply_to.chat_id,
                message_id=reply_to.message_id
            )
        except Exception as e:
            print(f"Forward to {target_id} error:", e)

    # Forward ke subscribers
    for sub_id in list(subscribers):
        try:
            await context.bot.forward_message(
                chat_id=sub_id,
                from_chat_id=reply_to.chat_id,
                message_id=reply_to.message_id
            )
            time.sleep(0.5 + random.random() * 0.3)
        except Exception as e:
            print(f"Subscriber {sub_id} removed due to error:", e)
            subscribers.discard(sub_id)

    # Hapus command
    try:
        await update.message.delete()
    except Exception:
        pass

# ================= UNSUBSCRIBE =================
async def unsub(update: Update, context: ContextTypes.DEFAULT_TYPE):
    subscribers.discard(update.effective_user.id)
    await update.message.reply_text("✅ Anda telah berhenti langganan.")

# ================= AUTO DELETE BOT MESSAGE =================
async def auto_delete(update: Update, context: ContextTypes.DEFAULT_TYPE):
    bot_id = context.bot.id
    if update.effective_chat.id == SOURCE_CHAT_ID and update.effective_user.id == bot_id:
        time.sleep(AUTO_DELETE_DELAY)
        try:
            await update.message.delete()
        except Exception:
            pass

# ================= MAIN BOT =================
app = ApplicationBuilder().token(BOT_TOKEN).build()

# Command handlers
app.add_handler(CommandHandler("start", start))
app.add_handler(CommandHandler(["menu", "help", "about", "profile", "contact"], start))
app.add_handler(CommandHandler("forward", forward))
app.add_handler(CommandHandler("unsub", unsub))

# Message handlers
app.add_handler(MessageHandler(filters.TEXT & (~filters.COMMAND), menu_handler))
app.add_handler(MessageHandler(filters.ALL, auto_delete))

# ================= FLASK KEEP ALIVE =================
flask_app = Flask(__name__)

@flask_app.route("/")
def index():
    return "🤖 Bot sedang berjalan"

def run_flask():
    flask_app.run(host="0.0.0.0", port=int(os.environ.get("PORT", 10000)))

threading.Thread(target=run_flask).start()

# ================= RUN BOT =================
print("✅ Bot sedang dijalankan...")
app.run_polling()

File: bot.js
// bot.js (versi lengkap, bahasa Malaysia, anti-link semua group + MongoDB subscriber)
import { Telegraf, Markup } from "telegraf";
import dotenv from "dotenv";
import { MongoClient } from "mongodb";
import express from "express";

dotenv.config();

// ================= KONFIGURASI ASAS =================
const BOT_TOKEN = process.env.BOT_TOKEN || "ISI_TOKEN_DI_SINI";
const ADMIN_USER_ID = 8146896736; // ID admin (akaun telegram anda)
const PORT = parseInt(process.env.PORT || "8080", 10);
// Wajib isi di .env: MONGODB_URI=mongodb+srv://botuser:PASSWORD@cluster0.uxxklgz.mongodb.net/botdb?retryWrites=true&w=majority
const MONGODB_URI = (process.env.MONGODB_URI || "").trim();

// ===== ID GROUP & CHANNEL =====
const SOURCE_CHAT_ID = -1002626291566; // GROUP UTAMA (tempat anda guna /forward)

const TARGET_CHAT_IDS = [
  // ===== GROUP LAIN =====
  -1003443785953,
  -1003355430208,
  -1003303586267,
  -1003351929392,
  -1003386119312,
  -1002068306604,
  -1002174638632,
  -1002112370494,
  -1002199080095,
  -1001925377693,
  -1002153443910,

  // ===== CHANNEL =====
  -1003175423118,
  -1003418215358,
  -1003410432304,
  -1003390131591,
  -1003379058057
];

const AUTO_DELETE_DELAY = 5000; // ms – auto delete mesej bot di group utama

// Tetapan siaran ke subscriber
const SUB_BATCH_SIZE = 20;            // hantar serentak ke 20 orang per batch
const SUB_DELAY_BETWEEN_BATCH = 1000; // jeda 1 saat antara batch

// Tetapan anti-spam
const ENABLE_LINK_ANTISPAM = true; // true = blok link dari bukan admin di semua group

// Senarai kata/frasa yang di-ban (semua dalam huruf kecil)
const BANNED_WORDS = [
  "new register",
  "free kredit luar",
  "bonus 100%",
  "kencing",
  "anjing"
].map(w => w.toLowerCase());

// ===== SEMAK BOT_TOKEN =====
if (!BOT_TOKEN || BOT_TOKEN === "ISI_TOKEN_DI_SINI") {
  console.error(
    "[STARTUP] ❌ BOT_TOKEN kosong / masih 'ISI_TOKEN_DI_SINI'. " +
    "Sila isi env BOT_TOKEN dahulu."
  );
  process.exit(1);
}

const bot = new Telegraf(BOT_TOKEN);

// Tangkap error telegraf supaya mudah debug
bot.catch((err, ctx) => {
  console.error("[TELEGRAF] Ralat pada update:", err.message, "update:", ctx.update);
});

// ================= MONGODB – SUBSCRIBER PERSISTENT =================
let mongoClient = null;
let subscribersCollection = null;
const DB_NAME = "botdb";
const SUBSCRIBERS_COLLECTION = "subscribers";

async function connectMongo() {
  if (!MONGODB_URI) {
    console.error(
      "[MONGODB] ❌ MONGODB_URI kosong. Sila isi dalam .env:\n" +
      "MONGODB_URI=mongodb+srv://botuser:PASSWORD@cluster0.uxxklgz.mongodb.net/botdb?retryWrites=true&w=majority"
    );
    return false;
  }
  const maxRetries = 3;
  const retryDelayMs = 2000;
  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      console.log("[MONGODB] Sambung... percubaan", attempt, "/", maxRetries);
      mongoClient = new MongoClient(MONGODB_URI);
      await mongoClient.connect();
      const db = mongoClient.db(DB_NAME);
      subscribersCollection = db.collection(SUBSCRIBERS_COLLECTION);
      await subscribersCollection.createIndex({ userId: 1 }, { unique: true });
      console.log("[MONGODB] ✅ Konek langsung ke MongoDB — koleksi:", SUBSCRIBERS_COLLECTION);
      return true;
    } catch (err) {
      console.error("[MONGODB] Percubaan", attempt, "gagal:", err.message);
      if (mongoClient) {
        try { await mongoClient.close(); } catch {}
        mongoClient = null;
        subscribersCollection = null;
      }
      if (attempt < maxRetries) {
        console.log("[MONGODB] Cuba lagi dalam", retryDelayMs / 1000, "saat...");
        await new Promise((r) => setTimeout(r, retryDelayMs));
      } else {
        console.error("[MONGODB] ❌ Gagal selepas", maxRetries, "percubaan. Sila semak MONGODB_URI dan Network Access di Atlas.");
        return false;
      }
    }
  }
  return false;
}

// Return true hanya bila subscriber BENAR-BENAR BARU (belum ada dalam DB)
async function addSubscriber(userId) {
  if (!subscribersCollection) return false;
  try {
    const result = await subscribersCollection.updateOne(
      { userId },
      { $set: { userId, updatedAt: new Date() } },
      { upsert: true }
    );
    return result.upsertedCount === 1;
  } catch (err) {
    console.error("[MONGODB] addSubscriber ralat:", err.message);
    return false;
  }
}

async function removeSubscriber(userId) {
  if (!subscribersCollection) return false;
  try {
    await subscribersCollection.deleteOne({ userId });
    return true;
  } catch (err) {
    console.error("[MONGODB] removeSubscriber ralat:", err.message);
    return false;
  }
}

async function getAllSubscribers() {
  if (!subscribersCollection) return [];
  try {
    const cursor = subscribersCollection.find({}, { projection: { userId: 1, _id: 0 } });
    const list = await cursor.toArray();
    return list.map((doc) => doc.userId);
  } catch (err) {
    console.error("[MONGODB] getAllSubscribers ralat:", err.message);
    return [];
  }
}

async function removeSubscribersByIds(userIds) {
  if (!subscribersCollection || !userIds.length) return;
  try {
    await subscribersCollection.deleteMany({ userId: { $in: [...userIds] } });
  } catch (err) {
    console.error("[MONGODB] removeSubscribersByIds ralat:", err.message);
  }
}

// Helper
const sleep = (ms) => new Promise((res) => setTimeout(res, ms));

// ================= EXPRESS (HEALTH CHECK) =================
const app = express();
app.get("/", (_, res) => res.send("🤖 Bot sedang berjalan"));
app.get("/health", (_, res) =>
  res.json({
    status: "ok",
    bot: "running",
    mongodb: subscribersCollection ? "connected" : "disconnected"
  })
);

// ================= /start & MENU UTAMA =================
async function sendStart(ctx) {
  const user = ctx.from || {};
  const username = user.username ? `@${user.username}` : user.first_name || "Bossku";

  // Daftar subscriber (simpan ke MongoDB)
  if (user.id) {
    const added = await addSubscriber(user.id);
    if (added) {
      try {
        await bot.telegram.sendMessage(
          ADMIN_USER_ID,
          `📌 Subscriber baru: ${username} (${user.id})`
        );
      } catch {}
    }
  }

  const inlineButtons = Markup.inlineKeyboard([
    [
      Markup.button.url("📢 Channel Utama", "https://t.me/afb88my"),
      Markup.button.url("💬 Group Cuci & Tips", "https://t.me/+b685QE242dMxOWE9")
    ],
    [
      Markup.button.url("🌐 Register & Login", "https://afb88my1.com/"),
      Markup.button.url("🎁 Group Hadiah", "https://t.me/Xamoi2688")
    ]
  ]);

  const replyKeyboard = Markup.keyboard([
    ["🌟 NEW REGISTER FREE 🌟"],
    ["📘 SHARE FACEBOOK 📘"],
    ["🔥 DAILY APPS FREE 🔥"],
    ["🌞 SOCIAL MEDIA 🌞"],
    ["🎉 TELEGRAM BONUS 🎉"]
  ]).resize();

  await ctx.replyWithAnimation(
    "https://media3.giphy.com/media/tXSLbuTIf37SjvE6QY/giphy.gif",
    {
      caption: `👋 Hi ${username} Bossku 😘

Kalau sudah join semua channel & group, amoi akan cuba bagi info paling untung untuk anda ❤️`,
      ...inlineButtons
    }
  );

  await ctx.reply("➤ Tekan /start bila-bila masa untuk kembali ke menu utama.", replyKeyboard);
}

bot.start(sendStart);
bot.command(["menu", "help", "about", "profile", "contact"], sendStart);

// ================= DATA MENU PRIVATE =================
const menuData = {
  "🌟 NEW REGISTER FREE 🌟": {
    url: "https://afb88my1.com/promotion",
    media: "https://ibb.co/BK2LVQ6t",
    caption: `🌟 NEW REGISTER BONUS AFB88 🌟
⚠️ Langgar syarat, semua point akan FORFEIT ⚠️
✅ Keperluan SLOT SAHAJA
✅ Free Credit R188
✅ Min WD/Cuci RM 6600
✅ Max Payment/WD RM20
✅ Dibenarkan main AFB GAMING (EVENT GAME SAHAJA)
✅ Dibenarkan main MEGAH5 | EPICWIN | PXPLAY2 | ACEWIN2 | RICH GAMING (EVENT GAME SAHAJA)
✅ Sila download apps untuk claim
📎 LINK: https://afb88.hfcapital.top/

⚠️ 1 nama 1 ID sahaja
⚠️ Nama daftar mesti sama dengan nama akaun bank

➤ CLICK /start BACK TO MENU`
  },
  "📘 SHARE FACEBOOK 📘": {
    url: "https://afb88my1.com/promotion",
    media: "https://ibb.co/Z6B55VcX",
    caption: `📘 PROMO SHARE FACEBOOK 📘
🧧 Free Credit RM68 🧧

Syarat:
✅ Join Telegram Channel
✅ Join Facebook Group
➡️ Share ke 5 group casino
➡️ Daily claim 1x sehari

➤ CLICK /start BACK TO MENU`
  },
  "🔥 DAILY APPS FREE 🔥": {
    url: "https://afb88my1.com/promotion",
    media: "https://ibb.co/nsmVQFbg",
    caption: `🔥 DAILY APPS FREE 🔥
🎁 Free Credit RM20
📌 Daily claim 1x
💰 Min WD RM600

➤ CLICK /start BACK TO MENU`
  },
  "🌞 SOCIAL MEDIA 🌞": {
    url: "https://afb88my1.com/promotion",
    media: "https://ibb.co/HfyD6DWw",
    caption: `🌞 SOCIAL MEDIA OFFICIAL AFB88 🌞
📘 Facebook
📸 Instagram
🎥 WhatsApp Group

➤ CLICK /start BACK TO MENU`
  },
  "🎉 TELEGRAM BONUS 🎉": {
    url: "https://afb88my1.com/promotion",
    media: "https://ibb.co/21qTqmtY",
    caption: `🎉 TELEGRAM BONUS KHAS 🎉
🎁 Free Credit RM30
✅ Claim 1x

➤ CLICK /start BACK TO MENU`
  }
};

// Data untuk butang link (reply keyboard)
const linkMenuData = {
  "📢 CHANNEL UTAMA": { url: "https://t.me/afb88my", label: "📢 BUKA CHANNEL" },
  "💬 GROUP CUCI & TIPS": { url: "https://t.me/+b685QE242dMxOWE9", label: "💬 BUKA GROUP" },
  "🌐 REGISTER & LOGIN": { url: "https://afb88my1.com/", label: "🌐 BUKA LAMAN" },
  "🎁 GROUP HADIAH AFB88": { url: "https://t.me/Xamoi2688", label: "🎁 BUKA GROUP" }
};

bot.hears(Object.keys(linkMenuData), async (ctx) => {
  if (ctx.chat?.type !== "private") return;
  const data = linkMenuData[ctx.message?.text];
  if (!data) return;
  await ctx.reply("Tekan butang di bawah untuk buka:", {
    ...Markup.inlineKeyboard([[Markup.button.url(data.label, data.url)]])
  });
});

bot.hears(Object.keys(menuData), async (ctx) => {
  if (ctx.chat?.type !== "private") return;

  const data = menuData[ctx.message?.text];
  if (!data) return;

  try {
    await ctx.replyWithPhoto(data.media, {
      caption: data.caption,
      ...Markup.inlineKeyboard([[Markup.button.url("CLAIM 🎁", data.url)]])
    });
  } catch (err) {
    console.error("Ralat hantar foto:", err.message);
    await ctx.reply(data.caption + `\n\n🔗 ${data.url}`, {
      ...Markup.inlineKeyboard([[Markup.button.url("CLAIM 🎁", data.url)]])
    });
  }
});

// ================= /unsub (berhenti langganan) =================
bot.command("unsub", async (ctx) => {
  if (!ctx.from) return;
  await removeSubscriber(ctx.from.id);
  await ctx.reply("✅ Anda telah berhenti langganan siaran dari bot ini.");
});

// ================= BANTUAN SIARAN / BROADCAST =================
let isBroadcastRunning = false;

// Forward ke group & channel target (bukan group utama)
async function broadcastToTargets(replyTo) {
  for (const targetId of TARGET_CHAT_IDS) {
    if (targetId === SOURCE_CHAT_ID) continue;

    try {
      await bot.telegram.forwardMessage(
        targetId,
        replyTo.chat.id,
        replyTo.message_id,
        { disable_notification: true }
      );
    } catch (err) {
      console.error("Ralat forward ke target", targetId, ":", err.message);
    }
  }
}

// Forward ke semua subscriber (baca dari MongoDB)
async function broadcastToSubscribers(replyTo) {
  const subscribers = await getAllSubscribers();
  if (!subscribers.length) return;

  const invalidIds = new Set();

  console.log(
    `[BROADCAST] Mula hantar ke ${subscribers.length} subscriber ` +
    `(batch=${SUB_BATCH_SIZE}, delay=${SUB_DELAY_BETWEEN_BATCH}ms)`
  );

  for (let i = 0; i < subscribers.length; i += SUB_BATCH_SIZE) {
    const batch = subscribers.slice(i, i + SUB_BATCH_SIZE);

    await Promise.all(
      batch.map(async (subId) => {
        try {
          await bot.telegram.forwardMessage(
            subId,
            replyTo.chat.id,
            replyTo.message_id,
            { disable_notification: true }
          );
        } catch (err) {
          console.error(
            "Ralat forward ke subscriber",
            subId,
            ":",
            err.description || err.message
          );

          const code = err?.response?.error_code;
          if (code === 400 || code === 403) {
            invalidIds.add(subId);
          }
        }
      })
    );

    if (i + SUB_BATCH_SIZE < subscribers.length) {
      await sleep(SUB_DELAY_BETWEEN_BATCH);
    }
  }

  if (invalidIds.size > 0) {
    await removeSubscribersByIds([...invalidIds]);
    console.log(`[BROADCAST] Buang ${invalidIds.size} subscriber tidak sah / block bot`);
  }

  console.log("[BROADCAST] Selesai hantar ke semua subscriber");
}

// ================= /forward (untuk admin di group utama) – tiada pesan, senyap =================
bot.command("forward", async (ctx) => {
  if (!ctx.from || ctx.from.id !== ADMIN_USER_ID) return;
  if (ctx.chat?.id !== SOURCE_CHAT_ID) return;

  const replyTo = ctx.message?.reply_to_message;
  if (!replyTo) return;

  if (isBroadcastRunning) return;

  isBroadcastRunning = true;

  try {
    await broadcastToTargets(replyTo);
    await broadcastToSubscribers(replyTo);
  } catch (err) {
    console.error("[BROADCAST] Ralat umum:", err);
  } finally {
    isBroadcastRunning = false;
    try { await ctx.deleteMessage(); } catch {}
  }
});

// ================= MODERASI: LINK & KATA TERLARANG – pesan amaran hanya bila melanggar =================
async function handleModeration(ctx) {
  if (!ENABLE_LINK_ANTISPAM) return;
  if (!ctx.chat) return;

  const chatType = ctx.chat.type;
  if (chatType !== "group" && chatType !== "supergroup") return;

  if (!ctx.from) return;

  const msg = ctx.message;
  if (!msg) return;

  const text = (msg.text || msg.caption || "").toString();
  const textLower = text.toLowerCase();
  const entities = msg.entities || msg.caption_entities || [];

  let hasLink = false;

  if (entities && entities.length) {
    if (entities.some(e => e.type === "url" || e.type === "text_link")) {
      hasLink = true;
    }
  }
  if (!hasLink && /https?:\/\/|www\.|t\.me\//i.test(text)) {
    hasLink = true;
  }

  const hasBannedWord = BANNED_WORDS.some(w => w && textLower.includes(w));

  if (!hasLink && !hasBannedWord) return;

  // Pengecualian: pemilik bot (ADMIN_USER_ID) sentiasa dibenarkan
  if (ctx.from.id === ADMIN_USER_ID) return;

  let isAdmin = false;
  try {
    const member = await ctx.getChatMember(ctx.from.id);
    if (member.status === "administrator" || member.status === "creator") {
      isAdmin = true;
    }
  } catch (err) {
    console.error("Gagal semak status ahli:", err.message);
  }

  if (isAdmin) return;

  try {
    await ctx.deleteMessage();
  } catch (err) {
    console.error("Gagal padam mesej melanggar peraturan:", err.message);
  }

  try {
    const warn = await ctx.reply(
      "⚠️ Mesej anda melanggar peraturan group (link luar / kata yang tidak dibenarkan). " +
      "Hanya admin dibenarkan kongsi link atau promo luar."
    );
    setTimeout(() => {
      bot.telegram.deleteMessage(warn.chat.id, warn.message_id).catch(() => {});
    }, 5000);
  } catch {}
}

bot.on("message", async (ctx) => {
  try {
    await handleModeration(ctx);
  } catch (err) {
    console.error("Ralat di handleModeration:", err.message);
  }

  const botId = bot.botInfo?.id;
  if (!botId) return;
  if (ctx.chat?.id === SOURCE_CHAT_ID && ctx.from?.id === botId) {
    const chatId = ctx.chat.id;
    const msgId = ctx.message?.message_id;
    if (chatId && msgId) {
      setTimeout(() => {
        bot.telegram.deleteMessage(chatId, msgId).catch(() => {});
      }, AUTO_DELETE_DELAY);
    }
  }
});

// ================= STARTUP =================
async function main() {
  console.log(
    "[STARTUP] PORT=" + PORT +
    ", BOT_TOKEN=" + (BOT_TOKEN ? "***ada***" : "KOSONG!") +
    ", MONGODB_URI=" + (MONGODB_URI ? "***ada***" : "KOSONG")
  );

  const mongoOk = await connectMongo();
  if (!mongoOk) {
    console.error("[STARTUP] ❌ Bot memerlukan MongoDB. Isi MONGODB_URI dalam .env dan pastikan Atlas Network Access dibenarkan.");
    process.exit(1);
  }

  const server = app.listen(PORT, "0.0.0.0", () => {
    console.log("[STARTUP] ✅ Server sedang mendengar pada port " + PORT);
  });

  server.on("error", (err) => {
    console.error("[STARTUP] ❌ Ralat Express:", err.message);
    process.exit(1);
  });

  try {
    const me = await bot.telegram.getMe();
    console.log(
      "[STARTUP] ✅ Bot berjaya sambung sebagai @" + me.username + " (id=" + me.id + ")"
    );
  } catch (err) {
    console.error(
      "[STARTUP] ❌ Gagal sambung ke Telegram. Kemungkinan besar BOT_TOKEN salah / bot telah dipadam."
    );
    console.error("[STARTUP] Butiran ralat:", err.message);
    return;
  }

  try {
    await bot.launch();
    console.log("[STARTUP] ✅ Bot Telegram sedang berjalan");
  } catch (err) {
    console.error("[STARTUP] ❌ Gagal mula bot:", err.message);
  }

  const stop = async () => {
    try { bot.stop("SIGTERM"); } catch {}
    if (mongoClient) {
      try { await mongoClient.close(); } catch {}
    }
    server.close(() => process.exit(0));
  };
  process.once("SIGINT", stop);
  process.once("SIGTERM", stop);
}

main().catch((err) => {
  console.error("[STARTUP] ❌ Ralat fatal:", err);
  process.exit(1);
});
